src/query/storage/m3/config.go

```diff
@@ -313,3 +316,77 @@ func (c ClustersStaticConfiguration) NewClusters(
 	return NewClusters(unaggregatedClusterNamespace,
 		aggregatedClusterNamespaces...)
 }
+
+// NewDynamicClusters instantiates a new Clusters instance that pulls
+// cluster information from etcd.
+func (c ClustersStaticConfiguration) NewDynamicClusters(
+	instrumentOpts instrument.Options,
+	opts ClustersStaticConfigurationOptions,
+) (Clusters, error) {
+	clients := make([]client.Client, 0, len(c))
+	for _, clusterCfg := range c {
+		clnt, err := clusterCfg.newClient(client.ConfigurationParameters{
+			InstrumentOptions: instrumentOpts,
+		}, opts.CustomAdminOptions...)
+		if err != nil {
+			return nil, err
+		}
+		clients = append(clients, clnt)
+	}
+
+	// Connect to all clusters in parallel
+	var (
+		wg   sync.WaitGroup
+		cfgs = make([]DynamicClusterNamespaceConfiguration, len(clients))
+
+		errLock  sync.Mutex
+		multiErr xerrors.MultiError
+	)
+	for i, clnt := range clients {
+		i := i
+		clnt := clnt
+		nsInit := clnt.Options().NamespaceInitializer()
+
+		// TODO(nate): move this validation to client.Options once static configuration of namespaces
+		// is no longer allowed.
+		if nsInit == nil {
+			return nil, errNoNamespaceInitializerSet
+		}
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			cfg := cfgs[i]
+			cfg.nsInitializer = nsInit
+			if opts.ProvidedSession != nil {
+				cfg.session = opts.ProvidedSession
+			} else if !opts.AsyncSessions {
+				var err error
+				cfg.session, err = clnt.DefaultSession()
+				if err != nil {
+					errLock.Lock()
+					multiErr = multiErr.Add(err)
+					errLock.Unlock()
+				}
+			} else {
+				cfg.session = m3db.NewAsyncSession(func() (client.Client, error) {
+					return clnt, nil
+				}, nil)
+			}
+		}()
+	}
+
+	// Wait
+	wg.Wait()
+
+	if !multiErr.Empty() {
+		return nil, multiErr.FinalError()
```

Comment:
Do these cfgs need any cleanup?

Reply:
Good catch. Probably smart to try and close the session of any configs that did get created

Reply:
It probably isn't that bad since this would just lead to a startup error right?

Reply:
That's correct. Would just fail to startup. Still, no harm and easy for me to add, tbh.